### PR TITLE
slog: faster and less memory-consumption

### PR DESCRIPTION
**Pull request by holiman**
Some previous discussion here: https://github.com/jwasinger/go-ethereum/pull/8 

----

These changes improves the performance of the non-coloured terminal formatting, _quite a lot_. 

```
name               old time/op    new time/op    delta
TerminalHandler-8    10.2µs ±15%     5.4µs ± 9%  -47.02%  (p=0.008 n=5+5)

name               old alloc/op   new alloc/op   delta
TerminalHandler-8    2.17kB ± 0%    0.40kB ± 0%  -81.46%  (p=0.008 n=5+5)

name               old allocs/op  new allocs/op  delta
TerminalHandler-8      33.0 ± 0%       5.0 ± 0%  -84.85%  (p=0.008 n=5+5)
```

I tried to _somewhat_ organize the commits, but the it might still be a bit chaotic. Some core insights: 

- The function `terminalHandler.Handl` uses a mutex, and writes all output immediately to 'upstream'. Thus, it can reuse a scratch-buffer every time. 
- This buffer can be propagated internally, making all the internal formatters either write directly to it,
- OR, make  use of the `tmp := buf.AvailableBuffer()` in some cases, where a byte buffer "extra capacity" can be temporarily used. 
- The `slog` package  uses `Attr` by value. It makes sense to minimize operating on them, since iterating / collecting into a new slice, iterating again etc causes copy-on-heap. Better to operate on them only once. 
- If we want to do padding, it's better to copy from a constant `space`-buffer than to invoke `bytes.Repeat` every single time. 
---

Todo: 

- [x]  `bytes.AvailableBuffer` was introduced in `go1.21` use something else for now, instead.

### CODE REVIEW

File: log/format.go [hunk 1]
```diff
@@ -284,54 +290,48 @@ func formatLogfmtBigInt(n *big.Int) string {
...
 			break
 		}
 	}
-	if !needsQuoting {
-		return s
+	if needsEscaping {
+		return strconv.AppendQuote(dst, s)
 	}
-	return strconv.Quote(s)
+	// No escaping needed, but we might have to place within quote-marks, in case
```
Comment (jwasinger) [inline log/format.go:327]: What is the reason for the distinction between `needsQuoting` and `needsEscaping`?

Comment (holiman) [reply log/format.go:327]: Good question!

So, if we need to `quote` it, all we need to do is put `"`-characters around the buffer. This is pretty cheap, we don't need to re-scan it, just write the things to the output. 

If we need to `escape` it, we need to place escapes before certain characters.  This requires re-scanning it again and is pretty costly.

Comment (holiman) [reply log/format.go:327]: And also, for the most part, we've simply added a key with a space it it, which needs quoting. Most keys/values doesn't actually require escaping.

**Review by jwasinger — APPROVED** [review]
Other than a few commented lines of code throughout the PR, LGTM

**Pull request merged**